sys/stl-compatibility/condition_variable.cpp

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (C) 2015 Hamburg University of Applied Sciences (HAW)
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup stl-compatibility
+ * @{
+ *
+ * @file    condition_variable.cpp
+ * @brief   STL condition variable drop in replacement
+ *
+ * @author  Raphael Hiesgen <raphael.hiesgen (at) haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include <cstdio>
+#include <stdexcept>
+
+#include "irq.h"
+#include "sched.h"
+#include "vtimer.h"
+#include "priority_queue.h"
+
+#include "riot/condition_variable.hpp"
+
+using namespace std::chrono;
+
+namespace riot {
+
+condition_variable::~condition_variable() { m_queue.first = NULL; }
+
+void condition_variable::notify_one() noexcept {
+  unsigned old_state = disableIRQ();
+  priority_queue_node_t* head = priority_queue_remove_head(&m_queue);
+  int other_prio = -1;
+  if (head != NULL) {
+    tcb_t* other_thread = (tcb_t*)sched_threads[head->data];
+    if (other_thread) {
+      other_prio = other_thread->priority;
+      sched_set_status(other_thread, STATUS_PENDING);
+    }
+    head->data = -1u;
+  }
+  restoreIRQ(old_state);
+  if (other_prio >= 0) {
+    sched_switch(other_prio);
+  }
+}
+
+void condition_variable::notify_all() noexcept {
+  unsigned old_state = disableIRQ();
+  int other_prio = -1;
+  while (true) {
+    priority_queue_node_t* head = priority_queue_remove_head(&m_queue);
+    if (head == NULL) {
+      break;
+    }
+    tcb_t* other_thread = (tcb_t*)sched_threads[head->data];
+    if (other_thread) {
+      auto max_prio
+        = [](int a, int b) { return (a < 0) ? b : ((a < b) ? a : b); };
+      other_prio = max_prio(other_prio, other_thread->priority);
+      sched_set_status(other_thread, STATUS_PENDING);
+    }
+    head->data = -1u;
+  }
+  restoreIRQ(old_state);
+  if (other_prio >= 0) {
+    sched_switch(other_prio);
+  }
+}
+
+void condition_variable::wait(unique_lock<mutex>& lock) noexcept {
+  if (!lock.owns_lock()) {
+    throw std::runtime_error("Mutex not locked.");
+  }
+  priority_queue_node_t n;
+  n.priority = sched_active_thread->priority;
+  n.data = sched_active_pid;
+  n.next = NULL;
+  // the signaling thread may not hold the mutex, the queue is not thread safe
```

Comment:
[This](https://github.com/actor-framework/actor-framework/blob/master/HACKING.md) is the coding style we use for C++ in CAF. We also have a file to use with clangformat.

In RIOT it would probably be preferable to use astyle for C++ as well?